mumd/src/network/udp.rs

```diff
@@ -228,31 +226,70 @@ pub async fn handle_pings(
         .await
         .expect("Failed to bind UDP socket");
 
-    let pending = Rc::new(Mutex::new(HashMap::new()));
+    let pending = Arc::new(Mutex::new(HashMap::new()));
 
-    let sender_handle = async {
+    let sender = async {
         while let Some((id, socket_addr, handle)) = ping_request_receiver.recv().await {
+            debug!("Sending ping {} to {}", id, socket_addr);
             let packet = PingPacket { id };
             let packet: [u8; 12] = packet.into();
             udp_socket.send_to(&packet, &socket_addr).await.unwrap();
-            pending.lock().await.insert(id, handle);
+            let (tx, rx) = oneshot::channel();
+            match pending.lock().await.entry(id) {
+                Entry::Occupied(_) => {
+                    warn!("Tried to send duplicate ping {}", id);
+                    continue;
+                }
+                Entry::Vacant(v) => {
+                    v.insert(tx);
+                }
+            }
+
+            tokio::spawn(async move {
+                let rx = rx.fuse();
+                let timeout = async {
+                    tokio::time::sleep(tokio::time::Duration::from_millis(1000)).await;
+                };
+                handle(select! {
+                    r = rx => match r {
+                        Ok(r) => Some(r),
+                        Err(_) => {
+                            warn!("Ping response sender dropped");
+                            None
+                        }
+                    },
+                    _ = timeout => None,
+                });
```

Comment:
This could be simplified a lot by the use of `tokio::time::Timeout` (and some clever use of functions to flatten results and options).